### ISSUE TITLE
Add heat bumps for crystals

### ISSUE THREAD

**Issue by mseaberg** [mseaberg/lcls_beamline_toolbox#19]
Need to think about the best way to do this. Phase error for wavefront (or add slope error instead by calculating diffraction angle and integrating?) plus consider angle of incidence change for crystal reflection intensity.

**Comment by mseaberg** [comment]
Done for 1D case. Might want to do 2D case at some point.

**Comment by mseaberg** [comment]
I still need to consider handling terms higher than second order for the diffraction direction. Another possible way to handle this is by doing the same thing I'm doing with the Grating case, where the height error makes a direct impact on the wavefront, rather than calculating the slope error and integrating. I would still need to calculate the slope error in order to calculate the crystal reflectivity. I can try this out and see if the behavior is the same as the current method for a single crystal.

**Comment by mseaberg** [comment]
Actually, better to just calculate the slope error of the beam based on the wavefront.

**Comment by mseaberg** [comment]
Made a first pass at this. It seems like it might be working...

**Comment by mseaberg** [comment]
For 1D heat bumps for now I think I will try doing it purely with numpy gradient and trapz methods, to see if that works. The phase is not well-behaved towards the edges which makes the global approach not work well. I could also try masking out the weak parts of the beam when the fit is performed, this might help smooth things out. As far as I can tell there is nothing that says the coordinates need to be equally spaced.

**Comment by mseaberg** [comment]
Finally I separated this into two parts:
1. First, the phase of the outgoing wavefront is handled as before, by only considering the contributions from the surface deformation, and the quadratic and linear terms of the phasefront.
2. In order to calculate the complex reflectivity of the crystal, the slope error and k_i, k_f are calculated a second time, this time by adding the gradient of the higher order phase of the incoming wavefront.
This has been checked in two ways. First, it exhibits the same behavior as the code that doesn't account for this, in the case where there are no thermal deformations of the crystal. Second, I have examined a particular high order crystal reflection and checked that the reflectivity as a function of slope error makes sense.

For now, I'm not going to put this into the master branch but instead into the new beta branch.
